### ISSUE TITLE
fix(markdown): improve handling of invalid markdown

### ISSUE THREAD

**Issue by nokome** [stencila/stencila#2405]
### What happened?

When Markdown (of any flavour is parsed) with the `strict: true` option (e.g. when parsed on the fly from a VSCode editor), we first check for unbalanced, colon, backtick, and dollar fences. An unbalanced fences emit an error (which appear as an error "diagnostic" is VSCode editor) and aborts further processing of the document. In addition, the `check` function emits warnings if there are an unbalanced number of characters in fences.

We do this so that unbalanced fences do not cause in-memory document nodes from being clobbered (and loosing properties such as execution outputs and provenance). For example, in this document if processing was not aborted, then the unbalanced backtick fence on line 6 would cause all subsequent Markdown in the document to be treated as Python code.

![image](https://github.com/user-attachments/assets/d069c5a6-1449-4114-a834-8b33de3c586f)

We currently show these `check` errors as diagnostics (the red lines above) so that the user has some indication as to why the current Markdown in the editor is out-of-sync with the preview panel. (In addition, we should prevent commands from being run when the Markdown is invalid (see https://github.com/stencila/stencila/issues/2388#issuecomment-2453627127).) There is currently a 1 second delay, in the publishing of these diagnostics:

https://github.com/stencila/stencila/blob/48ecf306b8eedc0cc9acb4670a7b77b60f29eb19/rust/lsp/src/text_document.rs#L502-L506

However, some feedback suggests that users might get a little freaked out by these error diagnostics. Some suggestions, not mutually exclusive, for how we might address this:

1. increase the above delay, say to 3 seconds
2. make the delay configurable with a default of say 3 seconds
3. reduce the diagnostic from `Error` level to `Warning` level
4. do not publish the diagnostics and leave the user to work out what is unbalanced
5. instead of in-editor diagnostics, pop up a message in the bottom right letting the user know that the Markdown is invalid (giving line and reason) and is out-of-sync with preview
6. pop up a message if the user attempts to run a command on an out-of-sync document, asking them to fix the error (giving line and reason).

### Which product?

Stencila VSCode extension

### Which version?

main branch

**Comment by jduckles** [comment]
I still don't like the "error message by default" we have set here. So I went and tried some other tools to see what they're doing: 
- Rstudio
- Quarto VSCode extension

Admittedly their execution model is write, then run, whereas Stencila's is live-updating. 

Neither of them say anything about an unbalanced fence at the time you type it.

Probably the closest analog to Stencila is RStudio (in "visual" mode, kind of a WSYWIG that is markdown enabled). Once the fence is typed along with a language and a new line is entered, it switches to a completed / balanced fence block:

https://github.com/user-attachments/assets/252cd997-f280-4d90-99e0-393a8279162e

Quarto VSCode extension quietly doesn't care at all about unbalanced fences. 

So based on this review of other tools, I think I prefer 4 or a light-weight (not heavy-handed) 6.

**Comment by brettc** [comment]
I've only been using this version for a week or so, but I found the warnings really useful. They helped me figure out what was expected. So maybe best to make them configurable. I think perhaps 3 + the ability to turn them off entirely.

One thing that would make it more obvious *without* having a warning or error is to use syntax highlight to define the fencing. At least in my setup, quarto has the code fences with a difference background color, but stencila markdown does not. 

Colorizing the `:::` is a harder proposition, but there are ways to do it (i.e. https://marketplace.visualstudio.com/items?itemName=MadsKristensen.RainbowBraces).

**Comment by kusalekanayake** [comment]
I'm finding myself having a love-hate relationship with them. On one hand they did help me learn the conventions quickly however I am not a fan of the error by default approach. 

> 1. increase the above delay, say to 3 seconds
> 2. make the delay configurable with a default of say 3 seconds

Extending the period before error is one way of making it a little less aggressive for sure. Is there some way to have it error only after a few seconds once the user has finished typing/once nor more input is detected? I.e. not as soon as the fence is setup.

> 5. instead of in-editor diagnostics, pop up a message in the bottom right letting the user know that the Markdown is invalid (giving line and reason) and is out-of-sync with preview
> 6. pop up a message if the user attempts to run a command on an out-of-sync document, asking them to fix the error (giving line and reason).

Having an error pop up on the lower right makes sense if they were to run it but makes less sense to me if it's a live report of the state of the document. 

> 3. reduce the diagnostic from Error level to Warning level

If we do nothing at all here, I am at least a fan of this, and then maybe 4. or 2.

**Comment by nokome** [comment]
> Is there some way to have it error only after a few seconds once the user has finished typing/once nor more input is detected? I.e. not as soon as the fence is setup.

This is how it currently works. There is a 500ms debounce delay for parsing the document so that it does not happen on every keypress, only when the user pauses for 0.5s or more.

Thanks for the opinions. Based on those, I think the solution is to implement the following:

>     3. reduce the diagnostic from `Error` level to `Warning` level

>     2. make the delay configurable with a default of say 3 seconds [including never displaying them]

>     6. pop up a message if the user attempts to run a command on an out-of-sync document, asking them to fix the error (giving line and reason).

In addition, implementing #2248 should help users understand which fences are balanced or not (in addition to providing an improved insight into the structure of the document).

**Comment by jduckles** [comment]
@kusalekanayake and I were wondering this morning, would it be possible to put the error as a red dot next to the line numbers? I see some linters do that, and when you mouse over them you can see the error in question.

**Comment by nokome** [comment]
Yep, that should be possible, and good suggestion!. Probably best implemented along side #2248 since it is related.
